alphadia/workflow/peptidecentric.py

```diff
@@ -365,11 +365,46 @@ def get_optimization_lock(self):
                 final_stop_index = stop_index  # final_stop_index is the number of elution groups that will be included in the calibration data
                 break
 
-        eg_idxes_for_calibration = self.elution_group_order[:final_stop_index]
-        return eg_idxes_for_calibration, precursor_df, fragments_df
+        optimization_lock_eg_idxes = self.elution_group_order[:final_stop_index]
 
-    #        self.eg_idxes_for_calibration = self.elution_group_order[:final_stop_index]
-    #        self.optimization_manager.fit({"classifier_version": self.fdr_manager.current_version})
+        optimization_lock_library_precursor_df = self.spectral_library._precursor_df[
+            self.spectral_library._precursor_df["elution_group_idx"].isin(
+                optimization_lock_eg_idxes
+            )
+        ]
+
+        optimization_lock_fragment_idxes = np.concatenate(
+            [
+                np.arange(row["flat_frag_start_idx"], row["flat_frag_stop_idx"])
+                for i, row in optimization_lock_library_precursor_df.iterrows()
+            ]
+        )
+
+        # Extract the fragments for the optimization lock and reset the indices to a consecutive range of positive integers. This simplifies future access based on position
+        optimization_lock_library_fragment_df = self.spectral_library._fragment_df.iloc[
+            optimization_lock_fragment_idxes
+        ].reset_index(drop=True)
+
+        # Change the fragment indices in the precursor_df to match the fragment indices in the optimization lock fragment_df instead of the full spectral library.
+        num_frags = (
+            optimization_lock_library_precursor_df["flat_frag_stop_idx"]
+            - optimization_lock_library_precursor_df["flat_frag_start_idx"]
+        )
+        optimization_lock_library_precursor_df["flat_frag_stop_idx"] = (
+            num_frags.cumsum()
+        )
+        optimization_lock_library_precursor_df["flat_frag_start_idx"] = (
+            optimization_lock_library_precursor_df[
+                "flat_frag_stop_idx"
+            ].shift(fill_value=0)
+        )
```

Comment:
could all this new code be extracted into a dedicated method?
also: would be really cool to have a simple unit test for these kinds of index manipulations ;-)

Reply:
I agee :)

Reply:
I have split this into three methods in the next PR (https://github.com/MannLabs/alphadia/pull/304#issue-2463350503); I've also introduced other changes so this would be awkward to merge if it was changed here as well.